### PR TITLE
[SofaBase] Package SofaBaseTopology

### PR DESCRIPTION
**Pull request by fredroy**
#1633

- Usual cleaning (pragma, namespaces)
- remove/reorganize useless includes
- fix some dependencies
- use only RGBAColor
- remove warnings
- use sofa::Size/sofa::Index (including missed functions, which makes this PR breaking for those inheriting for these components)

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
Compilation OK with SofaPython3 out-of-tree

File: SofaKernel/SofaBase/SofaBaseConfig.cmake.in [hunk 1]
```diff
@@ -4,25 +4,26 @@
...
 set(SOFABASE_TARGETS @SOFABASE_TARGETS@)
-set(SOFAEIGEN2SOLVER_HAVE_OPENMP @SOFAEIGEN2SOLVER_HAVE_OPENMP@)
 
 find_package(SofaFramework QUIET REQUIRED)
 find_package(SofaSimulation QUIET REQUIRED)
 
-# Eigen3 is required by SofaBaseTopology and SofaEigen2Solver
-find_package(Eigen3 QUIET REQUIRED)
+# Temporary list while doing packaging
+set(SOFABASE_MODULES SofaBaseLinearSolver SofaEigen2Solver SofaBaseTopology SofaBaseCollision SofaBaseUtils)
```
Comment (epernod) [inline SofaKernel/SofaBase/SofaBaseConfig.cmake.in:12]: Eigen3 is automatically added with SofaEigen2Solver?

File: SofaKernel/modules/SofaBaseTopology/CMakeLists.txt [hunk 2]
```diff
@@ -1,97 +1,110 @@
...
+    ${SOFABASETOPOLOGY_SRC}/TetrahedronSetGeometryAlgorithms.cpp
+    ${SOFABASETOPOLOGY_SRC}/TetrahedronSetTopologyContainer.cpp
+    ${SOFABASETOPOLOGY_SRC}/TetrahedronSetTopologyModifier.cpp
+    ${SOFABASETOPOLOGY_SRC}/TriangleSetGeometryAlgorithms.cpp
+    ${SOFABASETOPOLOGY_SRC}/TriangleSetTopologyContainer.cpp
+    ${SOFABASETOPOLOGY_SRC}/TriangleSetTopologyModifier.cpp
 )
 
+find_package(SofaSimulation REQUIRED)
 find_package(Eigen3 REQUIRED)
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/CMakeLists.txt:90]: eigen3 or eigen2Sofa ?

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/NumericalIntegrationDescriptor.cpp [hunk 3]
```diff
@@ -21,26 +21,16 @@
...
 
 
-template class  SOFA_BASE_TOPOLOGY_API NumericalIntegrationDescriptor<double,4>;
-template class  SOFA_BASE_TOPOLOGY_API NumericalIntegrationDescriptor<double,3>;
-template class  SOFA_BASE_TOPOLOGY_API NumericalIntegrationDescriptor<double,1>;
-
-
-
-} // namespace topology
+template class  SOFA_SOFABASETOPOLOGY_API NumericalIntegrationDescriptor<double,4>;
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/NumericalIntegrationDescriptor.cpp]: SReal ?

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/NumericalIntegrationDescriptor.h [hunk 4]
```diff
@@ -73,16 +66,10 @@ class NumericalIntegrationDescriptor {
 };
 
 #if  !defined(SOFA_COMPONENT_TOPOLOGY_NUMERICALINTEGRATIONDESCRIPTOR_CPP)
-extern template class SOFA_BASE_TOPOLOGY_API NumericalIntegrationDescriptor<double, 4>;
-extern template class SOFA_BASE_TOPOLOGY_API NumericalIntegrationDescriptor<double, 3>;
-extern template class SOFA_BASE_TOPOLOGY_API NumericalIntegrationDescriptor<double, 1>;
+extern template class SOFA_SOFABASETOPOLOGY_API NumericalIntegrationDescriptor<double, 4>;
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/NumericalIntegrationDescriptor.h]: same, SReal I would say...

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp [hunk 5]
```diff
@@ -19,36 +19,26 @@
 *                                                                             *
 * Contact information: contact@sofa-framework.org                             *
 ******************************************************************************/
-#include <sofa/simulation/Node.h>
 #include <SofaBaseTopology/SparseGridTopology.h>
-#include <sofa/core/visual/VisualParams.h>
+
+#include <sofa/core/behavior/MechanicalState.h>
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp:24]: BaseMechanicalState?

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp [hunk 6]
```diff
@@ -19,36 +19,26 @@
...
 #include <sofa/core/topology/BaseMeshTopology.h>
 #include <sofa/core/ObjectFactory.h>
 #include <sofa/helper/io/Mesh.h>
-#include <sofa/core/loader/MeshLoader.h>
-#include <sofa/helper/system/FileSystem.h>
 #include <sofa/helper/fixed_array.h>
 #include <sofa/helper/polygon_cube_intersection/polygon_cube_intersection.h>
-#include <sofa/helper/system/FileRepository.h>
-#include <sofa/defaulttype/VecTypes.h>
+#include <sofa/core/loader/VoxelLoader.h>
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp:30]: could you double check this one? Maybe BaseLoader?

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TopologyEngine.inl [hunk 7]
```diff
@@ -57,20 +51,13 @@ TopologyEngineImpl< VecT>::TopologyEngineImpl(t_topologicalData *_topologicalDat
 template <typename VecT>
 void TopologyEngineImpl< VecT>::init()
 {
-    // A pointData is by default child of positionSet Data
-    //this->linkToPointDataArray();  // already done while creating engine
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TopologyEngine.inl]: could you keep the comments.

**Comment by epernod** [comment]
Why is this one breaking if by default SofaBase are "auto loaded" ?

File: SofaKernel/SofaBase/SofaBaseConfig.cmake.in (hunk 1 above)
Comment (fredroy) [reply SofaKernel/SofaBase/SofaBaseConfig.cmake.in:12]: Yes

File: SofaKernel/modules/SofaBaseTopology/CMakeLists.txt (hunk 2 above)
Comment (fredroy) [reply SofaKernel/modules/SofaBaseTopology/CMakeLists.txt:90]: Eigen3 directly (see EdgeSetGeometryAlgorithms)

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/NumericalIntegrationDescriptor.h (hunk 4 above)
Comment (fredroy) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/NumericalIntegrationDescriptor.h]: I can set SReal yep

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp (hunk 5 above)
Comment (fredroy) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp:24]: Nope, line 589
```
        core::behavior::MechanicalState< sofa::defaulttype::Vec3Types > *mecha_temp =
                collisionTopology->getContext()->get< core::behavior::MechanicalState< sofa::defaulttype::Vec3Types > >();
```

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp (hunk 6 above)
Comment (fredroy) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp:30]: Already checked and it is needed:
```
void SparseGridTopology::buildAsFinest(  )
{
    VoxelLoader *loader;
    getContext()->get(loader);
    if( loader )
    {
        buildFromVoxelLoader(loader);
    }
    else
    ...
```

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TopologyEngine.inl (hunk 7 above)
Comment (fredroy) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TopologyEngine.inl]: ok

**Comment by fredroy** [comment]
> 
> 
> Why is this one breaking if by default SofaBase are "auto loaded" ?

Breaking because I changed some functions signatures (int to Size/Index)
Could be breaking for people overriding them

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp (hunk 6 above)
Comment (epernod) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/SparseGridTopology.cpp:30]: beurk

**Pull request merged**